### DISCUSSION TITLE
Suspense doesn't work on the client-side with SSR?

### DISCUSSION THREAD

**Discussion by affanshahid** [urql-graphql/urql#3141] [Q&A] [answered]
I am building a Next.js app and using next-urql. My understanding of using urql with SSR and Suspense is as follows: when `getInitialProps` is called on the server, urql will do its prepass logic to pre-fetch the data for the page render, however when `getInitialProps` is called on the client urql will not do a prepass, instead the page will render and since I have set `suspense: true` my usage of `useQuery` will trigger a suspension that I can then use with `<Suspense />`.

Is my understanding incorrect? because currently the prepass stuff works just fine but on the client-side no suspension takes place, instead the `fetching` state needs to be checked. 

My client configuration:

```ts
function urqlWrapperOptions(ssr: any, ctx?: NextPageContext): ClientOptions {
  const url = 'https://example.com';
  return {
    url: url,
    suspense: true,
    exchanges: [
      cacheExchange,
      ssr,
      fetchExchange,
    ],
    // ...
  };
}

```

**Comment by JoviDeCroock** [comment]
We disable prepass on the client-side as it's a server-side action, when we transition to a new useQuery on the client that needs a fetch we indeed stick to normal client-side rendering.

If you dislike or don't want this behavior you can customise this yourself by leveraging `getServerSideProps` instead

**Reply by affanshahid** [reply]
The "normal client-side rendering" does not support suspense? i.e it doesn't throw promises that I can catch using `<Suspense />`?

**Reply by JoviDeCroock** [reply, marked as answer]
It does, your explanation mentions that it is caught by `Suspense`, no? I could be misunderstanding your explanation but TLDR:

SSR: leverages prepass and rehydrates the data on the client during DOM-hydration
CSR (after DOM-hydration): leverages your normal client config i.e. throws promises when `suspense: true` 

Re-read your explanation and yes, we might have a bug in that case with overwriting suspense properties. Feel free to create a PR, the conditions to be aware of is that we should enable it during prepass, we should disable it during render-to-string and it should be the client setting on the client 😅

**Reply by affanshahid** [reply]
There is no bug, suspense was indeed working, I was just mis-applying the `<Suspense/>` boundary at the wrong place which was causing strange behaviour. After fixing that, the component does indeed suspend on the client as well. Thank you!
